fem/complex_fem.cpp

```diff
@@ -326,37 +361,105 @@ SesquilinearForm::FormLinearSystem(const Array<int> &ess_tdof_list,
    Vector b_i(&(b.GetData())[vsize], vsize);
    b_i *= s;
 
-   SparseMatrix * A_r = new SparseMatrix;
-   SparseMatrix * A_i = new SparseMatrix;
-   Vector X_0, B_0;
-
-   b_0 = b_r;
-   blfr->FormLinearSystem(ess_tdof_list, x_r, b_r, *A_r, X_0, B_0, ci);
+   int tvsize = fes->GetTrueVSize();
+   SparseMatrix * A_r = nullptr;
+   SparseMatrix * A_i = nullptr;
 
-   int tvsize = B_0.Size();
    X.SetSize(2 * tvsize);
    B.SetSize(2 * tvsize);
-   Vector X_r(X.GetData(), tvsize);
+
+   Vector X_0(tvsize), B_0(tvsize);
+   Vector X_r(X.GetData(),tvsize);
    Vector X_i(&(X.GetData())[tvsize], tvsize);
    Vector B_r(B.GetData(), tvsize);
    Vector B_i(&(B.GetData())[tvsize], tvsize);
-   X_r = X_0; B_r = B_0;
-
-   b_0 = 0.0;
-   blfi->FormLinearSystem(ess_tdof_list, x_i, b_0, *A_i, X_0, B_0, false);
-   B_r -= B_0;
-
-   b_0 = b_i;
-   blfr->FormLinearSystem(ess_tdof_list, x_i, b_0, *A_r, X_0, B_0, ci);
-   X_i = X_0; B_i = B_0;
+   if (RealInteg() && ImagInteg())
```

Comment:
There is some redundant code here that could be eliminated. There could be if statements for the real and imaginary cases separately, followed by the treatment of the real && imag cases, etc.

Reply:
You are right, just though it was more clear this way. I changed it in a way that the lines for the real part are not duplicated. For the imag part I kept the two cases separate, cause a lot of arguments that are passed to the FormLinearSystem method are different.